reconstruction/ecoli/dataclasses/process/metabolism_external_pathway.py

```diff
@@ -0,0 +1,258 @@
+                involved in the ODE
+            enzymeDict (dict): dictionary of enzyme counts
+            cellVolume (float): current volume of cell
+            nAvogadro (float): Avogadro's number
+            timeStepSec (float): current length of timestep in seconds
+            method (str): name of the ODE method to use
+            jit (bool): if True, use the jit compiled version of derivatives
+                functions
+
+
+        Returns:
+            moleculesNeeded (1d ndarray, ints): counts of molecules that need
+                to be consumed
+            allMoleculesChanges (1d ndarray, ints): expected changes in
+                molecule counts after timestep
+        """
+
+        subCounts, subNames, enzymeCounts, enzymeNames = [], [], [], []
+        # get the counts of the substrates.
+        # we assume that each reaction has only one substrate
+        for sub in self.substrates:
+            subCounts.append(moleculeDict[sub[0]])
+            subNames.append(sub[0])
+
+        # get the counts of the enzymes
+        # we assume that each reaction has only one enzyme
+        for reaction in self.rxn_ids:
+            # if the reaction has an enzyme
+            if self.enzymes[reaction]:
+                enzymeCounts.append(enzymeDict[self.enzymes[reaction][0]])
+                enzymeNames.append(self.enzymes[reaction][0])
+            #If the reaction is spontaneous, we set the enzyme counts to a high value to avoid creaitng a bottleneck.
+            #This might need to be modelled differently
+            else:
+                enzymeCounts.append(100000)
+                enzymeNames.append('None')
+
+
+        enzymeConc = np.array(enzymeCounts) / (cellVolume * nAvogadro)
+        sub_init = np.array(subCounts) / (cellVolume * nAvogadro)
+
+        moleculeCounts = list(moleculeDict.values())
+        y_init = np.array(moleculeCounts) / (cellVolume * nAvogadro)
+        derivatives = self.derivatives
+
+        sol = scipy.integrate.solve_ivp(
+            lambda t, y: derivatives([0, timeStepSec], sub_init, enzymeConc), [0, timeStepSec], y_init,
+            method=method, t_eval=[0, timeStepSec], atol=1e-8)
+        y = sol.y.T
+        y[y < 0] = 0
+        yMolecules = (y * (cellVolume * nAvogadro)) #.asNumber(COUNTS_UNITS/VOLUME_UNITS)
+
+        # Calculate changes in molecule counts for all molecules
+        allMoleculesChanges = yMolecules[-1, :] - yMolecules[0, :]
+
+        # Molecules needed are the change but only for the molecules corresponding to the negative terms in the stoich matrix
+        moleculesNeeded = np.negative(allMoleculesChanges).clip(min=0)
+
+        return moleculesNeeded, allMoleculesChanges, self.flux
+
+    def derivatives(self, t, y, enzymeC):
+        """
+        Calculate derivatives from stoichiometry and rates with argument order
+        for solve_ivp.
+        """
+        self.flux =  np.multiply(enzymeC, self._rates[0](y, t))
```

Comment:
It might be better to make this equation a bit more visible, since this is essentially the rate law equation that we're assuming for all of the reactions. I would suggest maybe pulling this out as a separate function that calculates the flux, and commenting it to make it clear that this function encodes the rate law equation.

Reply:
It would also be less confusing if `self.flux` is updated just once, using the initial values of `enzyme_conc` and `y`. It was a bit confusing to understand where the `self.flux` value was coming from, when it's hidden inside the `derivatives` function like this.